pkg/controller/operators/catalog/operator.go

```diff
@@ -474,6 +474,8 @@ func (o *Operator) syncSourceState(state grpc.SourceState) {
 	metrics.RegisterCatalogSourceState(state.Key.Name, state.Key.Namespace, state.State)
 
 	switch state.State {
+	case connectivity.Idle:
+		fallthrough
```

Comment:
Why are we falling through here exactly? 
Also, could you elaborate on what `Idle` connectivity state means?

Reply:
Good questions all around. So, in turn:
We need to bump the grpc_healthprobe version to satisfy the CVE -> this leads to a grpc library version bump.
The latest grpc library seems to have affected the connection state. Before the connection state would be READY once the CatalogSource pod was up. Now it sits on IDLE until there's an RPC call down the pipe. The transition from CONNECTION -> IDLE happens when there's "No RPC activity on channel for IDLE_TIMEOUT" [doc](https://gitlab.uni-hannover.de/tci-gateway-module/grpc/-/blob/972e31165218b49d93e5e1f1a1e8bbcd3fa830d1/doc/connectivity-semantics-and-api.md).

Let me refresh myself on the code base to give you a better explanation about the `passthrough`

Reply:
I see. `fallthrough` here (and in another place below) just means "if connectivity.Idle, do whatever you were supposed to do for connectivity.Ready". We probably want to confirm that that's the right thing to do to. 

i.e with this change in the library, when connectivity is in Idle, it feels like the intuitive thing to do would be for the controller to "sit idle" too, instead with "fallthrough" it's going to follow through with 1. Invalidate cache 2. Add to resolve queue...etc etc. Is that what we want? 

cc: @benluddy @njhale

Reply:
Also, this would be considered a breaking change for our consumers, so either we find a way to set the CatalogSource `.status.GRPCConnectionState.LastObservedState` to `Ready` for both `Idle` and `Ready`, or we need to send out communications to our community to alert them about the change.

Reply:
I spent yesterday afternoon poking around grpc and playing with [keepalive](https://github.com/grpc/grpc-go/blob/master/Documentation/keepalive.md) and changing both server and client settings. I couldn't really get it to return to the previous behavior. I think leaving it on IDLE could have the adverse effect of only finding out the CatalogSource is down when you actually try to talk to it. I removed the IDLE handling code above and did a test with a customer CatalogSource created in the default namespaces. I observed that the connection state stays on IDLE even if you kill the backing pod. Presumably because the content hadn't been cached yet (i.e. no gRPC calls). From the cluster administration perspective, I'm not sure that this is desirable. While we reduce network use and CatalogSource server resources improving scalability. On the other hand, ops would only get notified of a bad CatalogSource at the time of use. 

Another alternative, but it's a bit long winded:
Decouple the connection state from the registry server health (service can be up, but broken) and have metrics for connection state and registry server health implement the grpc health checks in the `RegistryReconciler`s ([exp branch](https://github.com/operator-framework/operator-lifecycle-manager/pull/2797))

For now, I've updated the code to create a connection in case of the IDLE state. This brings the behavior back inline with what we had before.